vervet-underground/internal/storage/mem/mem.go

```diff
@@ -6,81 +6,244 @@ package mem
+	revisions, ok := s.serviceVersionMappedRevisionSpecs[name][version]
+
 	if !ok {
 		return false, nil
 	}
-	_, ok = digests[digest]
+	_, ok = revisions[Digest(digest)]
 	return ok, nil
 }
 
 // NotifyVersion implements scraper.Storage.
-func (s *Storage) NotifyVersion(ctx context.Context, name string, version string, contents []byte, scrapeTime time.Time) error {
+func (s *MemoryStorage) NotifyVersion(ctx context.Context, name string, version string, contents []byte, scrapeTime time.Time) error {
 	s.mu.Lock()
 	defer s.mu.Unlock()
-	k := serviceVersion{service: name, version: version}
+
 	digest := storage.Digest(contents)
-	if digests, ok := s.contents[k]; ok {
-		if _, ok := digests[digest]; ok {
+	typedDigest := Digest(digest)
+
+	parsedVersion, err := vervet.ParseVersion(version)
+	if err != nil {
+		log.Error().Err(err).Msgf("Failed to resolve Vervet version for %s : %s", name, version)
+		return err
+	}
+
+	// Check if service and version structures are initialized
+	if _, ok := s.serviceVersionMappedRevisionSpecs[name]; !ok {
+		s.serviceVersionMappedRevisionSpecs[name] = VersionMappedRevisionSpecs{}
+	}
+
+	if revisions, ok := s.serviceVersionMappedRevisionSpecs[name][version]; ok {
+		if _, ok := revisions[typedDigest]; ok {
 			return nil
 		}
 	} else {
-		s.contents[k] = map[string][]byte{}
+		s.serviceVersionMappedRevisionSpecs[name][version] = MappedRevisionSpecs{}
+	}
+
+	revisions, ok := s.serviceVersionMappedRevisionSpecs[name][version]
+	if !ok {
+		err = errors.Errorf("Error reconciling current processing version in storage %s : %s", name, version)
+		log.Error().Err(err).Msgf("Could not save version %s : %s", name, version)
+		return err
+	}
+
+	// If the version is newly initialized, meaning no revisions exist,
+	// create the new vervet.VersionSlice with that version initialized
+	// else, append it to the existing service's VersionSlice
+	if len(revisions) == 0 {
+		if _, ok = s.serviceVersions[name]; !ok {
+			s.serviceVersions[name] = vervet.VersionSlice{*parsedVersion}
+		} else {
+			s.serviceVersions[name] = append(s.serviceVersions[name], *parsedVersion)
+			// sort versions when new ones are introduced to maintain BST functionality
+			sort.Sort(s.serviceVersions[name])
+		}
+	}
+	// End of initializations
+
+	// add the new ContentRevision
+	s.serviceVersionMappedRevisionSpecs[name][version][typedDigest] = ContentRevision{
+		serviceVersion: name + "_" + version,
+		timestamp:      scrapeTime,
+		digest:         typedDigest,
+		blob:           contents,
+	}
+
+	return nil
+}
+
+func (s *MemoryStorage) CollateVersions() error {
+	var errs error
+	s.gatherVersionsRequiredToCollate()
+	for _, version := range s.collatedVersions {
+		err := s.collateVersion(version)
+		if err != nil {
+			log.Error().Err(err).Msgf("Could not Collate Version %s", version.String())
+			errs = multierr.Append(errs, err)
+		}
+	}
+	return errs
+}
+
+func (s *MemoryStorage) GetCollatedVersionSpecs() CollatedVersionMappedSpecs {
```

Comment:
```suggestion
func (s *MemoryStorage) Versions() CollatedVersionMappedSpecs {
```

It may be fair to assume that the versions our storage provides are always collated, if that is its job to do that for us. So perhaps the name can be more concise?

Reply:
`Versions` would returning the `collatedVersions vervet.VersionSlice` value. These are the actual OAS.

Like `/openapi` vs. `/openapi/{version}`. It's the latter,  but for each one. I used it as testing function to validate my assumptions. It can be removed and is probably not permanent.

Reply:
Again, where naming too short is failing here.